request/request.go

```diff
@@ -0,0 +1,211 @@
+package request
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"reflect"
+	"time"
+)
+
+type URL struct {
+	scheme, host, path string
+	queries            url.Values
+}
+
+// NewURL instantiantes a new URL given an raw url (optional)
+func NewURL(rawURL ...string) *URL {
+	u := &URL{
+		queries: url.Values{},
+	}
+
+	if len(rawURL) == 0 || rawURL[0] == "" {
+		return u
+	}
+
+	parsedURL, err := url.Parse(rawURL[0])
+	if err != nil {
+		return u
+	}
+
+	return &URL{
+		scheme:  parsedURL.Scheme,
+		host:    parsedURL.Host,
+		path:    parsedURL.Path,
+		queries: parsedURL.Query(),
+	}
+}
+
+// SetScheme sets the URL scheme
+func (u *URL) SetScheme(s string) *URL {
+	if s != "" {
+		u.scheme = s
+	}
+	return u
+}
+
+// SetHost sets the URL host
+func (u *URL) SetHost(h string) *URL {
+	if h != "" {
+		u.host = h
+	}
+	return u
+}
+
+// SetPath sets the URL path
+func (u *URL) SetPath(p string) *URL {
+	if p != "" {
+		u.path = p
+	}
+	return u
+}
+
+// SetPath adds a new query to URL
+func (u *URL) AddQuery(key, val string) *URL {
+	if key != "" && val != "" {
+		u.queries.Set(key, val)
+	}
+	return u
+}
+
+// Parse parses the URL to a raw url string
+func (u *URL) Parse() (s string, err error) {
+	host := u.host
+	if u.scheme != "" {
+		host = fmt.Sprintf("%s://%s", u.scheme, host)
+	}
+
+	rawURL, err := url.JoinPath(host, u.path)
+	if err != nil {
+		err = fmt.Errorf("unable to parse url with host %s and path %s", host, u.path)
+		return
+	}
+
+	parsedURL, _ := url.Parse(rawURL)
+	parsedURL.RawQuery = u.queries.Encode()
+
+	s = parsedURL.String()
+	return
+}
+
+type Response http.Response
+
+// IsSuccessCode checks if response status code is from a success response.
+func (r *Response) IsSuccessCode() bool {
+	return r.StatusCode >= 200 && r.StatusCode <= 299
+}
+
+// IsFailureCode checks if response status code is from an error response.
+func (r *Response) IsFailureCode() bool {
+	return r.StatusCode >= 400 && r.StatusCode <= 599
+}
+
+// DecodeJSON decodes the response body into an object pointer
+func (r *Response) DecodeJSON(value any) error {
+	if value == nil || reflect.ValueOf(value).Kind() != reflect.Ptr {
+		return errors.New("value must be a valid pointer of a struct")
+	}
```

Comment:
Essa validação é desnecessária, o `Decode` já se encarrega disso 😉 

Além disso, podemos mandar um map ou um slice pra essa função.
maps e slices são valores referenciais e não precisam de ponteiros para fazer Decode.

Reply:
Não acho que a validação seja desnecessária, pelo menos não por esse motivo.

Considerando a forma que foi implementado, recebendo um parâmetro do tipo `any`, o decode só funciona se esse valor for um ponteiro. Caso contrário, é lançado esse erro: `json: Unmarshal(non-pointer)`.

Por outro lado, se eu decidir fazer algo do tipo:

```golang
func (r *Response) DecodeJSON(value any) error {
	decoder := json.NewDecoder(r.Body)
	return decoder.Decode(&value)
}
```

O decode também não irá funcionar corretamente. O motivo real eu não sei com exatidão (suspeito que seja devido ao parâmetro ser do tipo any), mas o valor não é decodificado corretamente. Eu fiz testes com structs, array de structs e com maps, e simplesmente o valor não é setado corretamente nos parâmetros enviados.

Caso queira, posso te mostrar os testes que fiz. Ou então, se tiver alguma outra sugestão que seja mais apropriada, podemos adicionar lá 😉

Reply:
saquei! tranquilo